app/app.go

```diff
@@ -484,11 +486,27 @@ func (app *App) EndBlocker(ctx sdk.Context, req abci.RequestEndBlock) abci.Respo
 			}
 		}
 		// from v2 to v3 and onwards we use a signaling mechanism
-	} else if shouldUpgrade, newVersion := app.SignalKeeper.ShouldUpgrade(ctx); shouldUpgrade {
-		// Version changes must be increasing. Downgrades are not permitted
-		if newVersion > currentVersion {
-			app.BaseApp.Logger().Info("upgrading app version", "current version", currentVersion, "new version", newVersion)
-			app.SetAppVersion(ctx, newVersion)
+	} else if shouldUpgrade, upgrade := app.SignalKeeper.ShouldUpgrade(ctx); shouldUpgrade {
+		// Version changes must be increasing. Downgrades are not permitted.
+		if upgrade.AppVersion > currentVersion {
+			app.BaseApp.Logger().Info("upgrading app version", "current version", currentVersion, "new version", upgrade.AppVersion)
+
+			if currentVersion == v3 { // v3 -> v4 needs to schedule an upgrade
+				plan := upgradetypes.Plan{
+					Name:   fmt.Sprintf("v%d", upgrade.AppVersion),
+					Height: upgrade.UpgradeHeight + 1, // next block executing the upgrade.
```

Comment:
[question] why is the + 1 needed here? Will the block at UpgradeHeight have app version 3 or 4 in it?

```suggestion
					Height: upgrade.UpgradeHeight + 1, // the block after the upgrade height executes the upgrade.
```

Reply:
Because of how your fork was working. The migrations were done in `Commit` within the same block, and the app version updated. For the v3 -> v4, and all successive migrations, the app version will be updated at `UpgradeHeight`, and the same block will be running the new binary containing the migrations.